File: packages/aid-go/discover.go
package aid

import (
    "context"
    "net"
    "strings"
    "time"

    "golang.org/x/net/idna"
)

// lookupTXT is an indirection to make unit testing easier.
var lookupTXT = net.DefaultResolver.LookupTXT

// Discover queries DNS for the _agent TXT record and parses it.
// Returns record and a TTL (0 when unknown).
// DiscoveryOptions provides optional behavior controls for discovery.
type DiscoveryOptions struct {
    // Protocol: when set, try _agent._<proto>.<domain> then _agent.<proto>.<domain> before base.
    Protocol string
    // WellKnownFallback: if true, attempt HTTPS .well-known fallback on ERR_NO_RECORD or ERR_DNS_LOOKUP_FAILED.
    WellKnownFallback bool
    // WellKnownTimeout: timeout for the .well-known HTTP fetch.
    WellKnownTimeout time.Duration
}

// Discover retains the original signature for backward compatibility.
// It performs DNS-first discovery and falls back to HTTPS .well-known.
func Discover(domain string, timeout time.Duration) (AidRecord, uint32, error) {
    opts := DiscoveryOptions{WellKnownFallback: true, WellKnownTimeout: 2 * time.Second}
    return DiscoverWithOptions(domain, timeout, opts)
}

// DiscoverWithOptions performs discovery with protocol-specific DNS flow and well-known controls.
func DiscoverWithOptions(domain string, timeout time.Duration, opts DiscoveryOptions) (AidRecord, uint32, error) {
    // IDN → A-label
    alabel, _ := idna.ToASCII(domain)

    // Helper to resolve a specific FQDN
    resolve := func(fqdn string) (AidRecord, uint32, error) {
        fqdn = strings.TrimSuffix(fqdn, ".")
        ctx, cancel := context.WithTimeout(context.Background(), timeout)
        defer cancel()
        txts, err := lookupTXT(ctx, fqdn)
        if err != nil {
            return AidRecord{}, 0, newAidError("ERR_DNS_LOOKUP_FAILED", err.Error())
        }
        var lastErr error
        for _, txt := range txts {
            rec, perr := Parse(txt)
            if perr == nil {
                // PKA handshake if present (redirects disabled in performPKAHandshake)
                if rec.Pka != "" {
                    if err := performPKAHandshake(rec.URI, rec.Pka, rec.Kid, timeout); err != nil {
                        return AidRecord{}, 0, err
                    }
                }
                return rec, 0, nil
            }
            lastErr = perr
        }
        if lastErr != nil {
            return AidRecord{}, 0, lastErr
        }
        return AidRecord{}, 0, newAidError("ERR_NO_RECORD", "No valid AID record in TXT answers")
    }

    // Query order
    var names []string
    if opts.Protocol != "" {
        names = append(names,
            DnsSubdomain+"._"+opts.Protocol+"."+alabel,
            DnsSubdomain+"."+opts.Protocol+"."+alabel,
        )
    }
    names = append(names, DnsSubdomain+"."+alabel)

    var lastErr *AidError
    for _, name := range names {
        rec, ttl, err := resolve(name)
        if err == nil {
            return rec, ttl, nil
        }
        if ae, ok := err.(*AidError); ok {
            lastErr = ae
            if ae.Symbol != "ERR_NO_RECORD" {
                // Only continue to next name on no-record; otherwise propagate
                break
            }
            continue
        }
        // Non-AidError: treat as DNS failure
        lastErr = newAidError("ERR_DNS_LOOKUP_FAILED", err.Error())
        break
    }

    // DNS failed → optionally fallback to well-known
    if opts.WellKnownFallback && lastErr != nil && (lastErr.Symbol == "ERR_NO_RECORD" || lastErr.Symbol == "ERR_DNS_LOOKUP_FAILED") {
        rec, werr := fetchWellKnown(alabel, firstNonZero(opts.WellKnownTimeout, 2*time.Second))
        if werr != nil {
            return AidRecord{}, 0, werr
        }
        if rec.Pka != "" {
            if err := performPKAHandshake(rec.URI, rec.Pka, rec.Kid, timeout); err != nil {
                return AidRecord{}, 0, err
            }
        }
        return rec, uint32(DnsTtlMin), nil
    }
    if lastErr != nil {
        return AidRecord{}, 0, lastErr
    }
    return AidRecord{}, 0, newAidError("ERR_DNS_LOOKUP_FAILED", "DNS query failed")
}

func firstNonZero(d time.Duration, def time.Duration) time.Duration { if d > 0 { return d }; return def }

File: packages/aid-rs/src/discover.rs
use std::time::Duration;

use crate::errors::AidError;
use crate::parser::parse;
use crate::record::AidRecord;
use hickory_resolver::TokioAsyncResolver;
use idna::domain_to_ascii;

#[cfg(feature = "handshake")]
use crate::pka::perform_pka_handshake;

#[cfg(feature = "handshake")]
use crate::well_known::fetch_well_known;

/// Discover an AID record for the given domain using DNS TXT at _agent.<domain>.
/// Falls back to HTTPS .well-known when DNS has no record or lookup fails.
pub async fn discover(domain: &str, timeout: Duration) -> Result<AidRecord, AidError> {
    let opts = DiscoveryOptions { protocol: None, timeout, well_known_fallback: true, well_known_timeout: Duration::from_secs(2) };
    discover_with_options(domain, opts).await
}

pub struct DiscoveryOptions {
    pub protocol: Option<String>,
    pub timeout: Duration,
    pub well_known_fallback: bool,
    pub well_known_timeout: Duration,
}

pub async fn discover_with_options(domain: &str, options: DiscoveryOptions) -> Result<AidRecord, AidError> {
    // IDNA → A-label
    let alabel = domain_to_ascii(domain).unwrap_or_else(|_| domain.to_string());
    let mut names: Vec<String> = Vec::new();
    if let Some(proto) = &options.protocol {
        names.push(format!("_agent._{}.{}", proto, alabel));
        names.push(format!("_agent.{}.{}", proto, alabel));
    }
    names.push(format!("_agent.{}", alabel));

    // DNS lookup using system resolver
    let resolver = TokioAsyncResolver::tokio_from_system_conf()
        .map_err(|e| AidError::new("ERR_DNS_LOOKUP_FAILED", e.to_string()))?;

    // iterate names
    let mut last_err: Option<AidError> = None;
    for name in names {
        let txt_lookup = tokio::time::timeout(options.timeout, resolver.txt_lookup(name.clone())).await;
        match txt_lookup {
            Err(_) => {
                last_err = Some(AidError::new("ERR_DNS_LOOKUP_FAILED", "DNS query timeout"));
                break;
            }
            Ok(Err(e)) => {
                let msg = e.to_string().to_lowercase();
                let code = if msg.contains("nxdomain") || msg.contains("no record") || msg.contains("no data") { "ERR_NO_RECORD" } else { "ERR_DNS_LOOKUP_FAILED" };
                let err = AidError::new(code, e.to_string());
                if code != "ERR_NO_RECORD" { last_err = Some(err); break; }
                last_err = Some(err);
                continue;
            }
            Ok(Ok(lookup)) => {
                for r in lookup.iter() {
                    let s = r.txt_data().iter().map(|b| String::from_utf8_lossy(b).to_string()).collect::<Vec<_>>().join("");
                    let raw = s.trim();
                    if raw.to_ascii_lowercase().starts_with("v=aid1") {
                        if let Ok(mut rec) = parse(raw) {
                            #[cfg(feature = "handshake")]
                            {
                                if let (Some(pka), Some(kid)) = (rec.pka.clone(), rec.kid.clone()) {
                                    perform_pka_handshake(&rec.uri, &pka, &kid, options.timeout).await?;
                                }
                            }
                            return Ok(rec);
                        }
                    }
                }
                last_err = Some(AidError::new("ERR_NO_RECORD", format!("No valid AID record found for {}", name)));
                continue;
            }
        }
    }

    // Fallback
    if options.well_known_fallback {
        #[cfg(feature = "handshake")]
        {
            return fetch_well_known(&alabel, options.well_known_timeout).await;
        }
    }
    Err(last_err.unwrap_or_else(|| AidError::new("ERR_DNS_LOOKUP_FAILED", "DNS query failed")))
}
